src/tools/tests/manifest2proto-test.ts

```diff
@@ -9,13 +9,259 @@
+          }
+        }
+      }
+    });
+  });
+
+  it('encodes reference type', async () => {
+    const reference = EntityType.make(['Foo'], {value: 'Text'}).referenceTo();
+    assert.deepStrictEqual(typeProto.fromObject(await typeToJSON(reference)).toJSON(), {
+      reference: {
+        referredType: {
+          entity: {
+            schema: {
+              names: ['Foo'],
+              fields: {
+                value: {
+                  primitive: 'TEXT'
+                }
+              },
+              hash: '1c9b8f8d51ff6e11235ac13bf0c5ca74c88537e0',
+            }
+          }
+        }
+      }
+    });
+  });
+
+  it('encodes schemas with primitives and collections of primitives', async () => {
+    const manifest = await Manifest.parse(`
+      particle Abc in 'a/b/c.js'
+        input: reads X Y Z {
+          a: Text,
+          b: Number,
+          c: Boolean,
+          d: [Text],
+          e: [Number]
+        }
+    `);
+    const fromProto = manifestProto.fromObject(await manifestToJSON(manifest)).toJSON();
+    const schema = fromProto.particleSpecs[0].connections[0].type.entity.schema;
+
+    assert.deepStrictEqual(schema.names, ['X', 'Y', 'Z']);
+    assert.deepStrictEqual(schema.fields, {
+      a: {primitive: 'TEXT'},
+      b: {primitive: 'NUMBER'},
+      c: {primitive: 'BOOLEAN'},
+      d: {collection: {collectionType: {primitive: 'TEXT'}}},
+      e: {collection: {collectionType: {primitive: 'NUMBER'}}}
+    });
+  });
+
+  it('encodes schemas with entity references', async () => {
+    const manifest = await Manifest.parse(`
+      particle Abc in 'a/b/c.js'
+        input: reads {
+          a: &Product {name: Text},
+          b: [&Review {rating: Number}]
+        }
+    `);
+    const fromProto = manifestProto.fromObject(await manifestToJSON(manifest)).toJSON();
+    const schema = fromProto.particleSpecs[0].connections[0].type.entity.schema;
+
+    assert.deepStrictEqual(schema.fields, {
+      a: {reference: {referredType: {entity: {schema: {
+        names: ['Product'],
+        fields: {
+          name: {primitive: 'TEXT'}
+        },
+        hash: 'a76bdd3a638fc17a5b3e023edb542c1e891c4c89'
+      }}}}},
+      b: {collection: {collectionType: {reference: {referredType: {entity: {schema: {
+        names: ['Review'],
+        fields: {
+          rating: {primitive: 'NUMBER'},
+        },
+        hash: '2d3317e5ef54fbdf3fbc02ed481c2472ebe9ba66'
+      }}}}}}},
+    });
+  });
+
+  // On the TypeScript side we serialize .arcs file and validate it equals the .pb.bin file.
+  // On the Kotlin side we deserialize .pb.bin and validate it equals deserialized .textproto file.
+  // This ensures that at least all the constructs used in the .arcs file can be serialized in TS
+  // and deserialized in Kotlin to the extent that they are present in the .textproto file.
+  it('encodes the Manifest2ProtoTest manifest', async () => {
+    // You can update the pb.bin file with the following command:
```

Comment:
Found it!

For bonus points, you could do something similar to what we do in schema-generator-test.ts and print out hte command to run when the test fails.

Reply:
Done!